### PR TITLE
Simplify stream creation for set of different classes

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change introduces `Streamable` interface that is similar to `Itreable` but also provides `stream()` and `size()` methods.
I believe it should also replace our usages of `AbstractCollection` to avoid exposing mutation methods such as add/remove that are implemented with default `throw new UnsupportedOpertaionException()`.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
  * This class holds all {@link DiscoveryNode} in the cluster and provides convenience methods to
  * access, modify merge / diff discovery nodes.
  */
-public class DiscoveryNodes extends AbstractCollection<DiscoveryNode> implements SimpleDiffable<DiscoveryNodes> {
+public class DiscoveryNodes implements Streamable<DiscoveryNode>, SimpleDiffable<DiscoveryNodes> {
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:43]: `DiscoveryNodes` is a good example here.

We are not using it with collections add/addAll/remove inherited from `AbstractCollection`, however there are valid stream/iterator usages of this class

File: server/src/main/java/org/elasticsearch/common/util/Streamable.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.common.util.iterable.Iterables;
+
+import java.util.stream.Stream;
+
+public interface Streamable<T> extends Iterable<T> {
+
+    default Stream<T> stream() {
+        return Iterables.stream(this);
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/common/util/Streamable.java:18]: I think we can use a sized spliterator here?

```java
StreamSupport.stream(Spliterators.spliterator(iterator(), size(), Spliterator.SIZED), false)
```

File: server/src/main/java/org/elasticsearch/common/util/Streamable.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    int size();
+
+    default boolean isEmpty() {
+        return size() == 0;
+    }
+}
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/common/util/Streamable.java:26]: I guess it would be nice to add a bridge method to `Collection` in order to adapt `Streamable` objects to the APIs that accept `Collection`

```java
default Collection<T> asCollection() {
        Iterator<T> iterator = iterator();
        int size = size();
        return new AbstractCollection<T>() {
            @Override
            public Iterator<T> iterator() {
                return iterator;
            }

            @Override
            public int size() {
                return size;
            }
        };
    }
```

**Review by original-brownbear — COMMENTED** [review]
I read through this for a little an I'm not sure I'm a fan to be honest.

To me it seems making this stuff `AbstractCollection` was the initial mistake not the fact that it exposes mutation operators.
None of the use cases even leverage the internals of the objects that return a stream. It's always just StreamSupport.stream(spliterator(), false);
-> why not just add a utility for that to `Iterables` and use that if we absolutely must use streams somewhere? It's less code to do so and we don't introduce a new unused (we never seem to use streamable in the abstract anywhere?) interface?

Bottom line: If we don't have any utilities using the interface then it's unused and we shouldn't introduce it. If we want to add utilities for anything around it, then those could just as well use `Iterable` and map to stream in the way I suggested above?

**Comment by idegtiarenko** [comment]
As discussed with @original-brownbear we should review the usages of stream to ensure this does not cause performance degradation critical places. Closing for now
